### PR TITLE
waypoints and portals

### PR DESCRIPTION
**Pull request by kostmo**
**Portals** are a prerequisite to subworlds (#144).

However, portals as a first-class feature have utility that is independent of subworlds.  The Hackman scenario (#1135), for example, currently has to provide redundant implementations of portal logic for the `base` robot and for system robots.  Furthermore, the locations specified in these bespoke portal implementations are hard-coded in the `.sw` file, so are brittle with changes to the map.

# Demos
    scripts/play.sh --scenario data/scenarios/Testing/1356-portals/automatic-waypoint-patrol.yaml
    
image

    scripts/play.sh --scenario scenarios/Testing/1356-portals/portals-flip-and-rotate.yaml --autoplay

image

# Portal and waypoint semantics

## Waypoints

**Waypoints** are introduced as an intermediate abstraction to better accommodate substructures.

A `waypoint` command, independent of portal mechanics, is also introduced to allow system robots to query named locations.

* **Waypoints** may be defined either in a standalone list (as a sibling field to the `map`), or within the `map` via the `palette`.
    * An **object notation** is now supported as an alternative to, and backwards compatible with, the **list notation** for palette values.
    * Using the palette to define waypoints is typically more convenient, as it allows specification of both a waypoint **and** an entity which can be used to depict the waypoint location.
* Waypoints can be defined at the root level or within "substructures".
* In contrast with "destructive" entity overlay from overlapping substructures, waypoints are aggregated from all substructures, even when overlapping.
* Multiple waypoints under different names can exist at the same location.  I.e. waypoints can have *plural multiplicity*.
    * Uniqueness of locations for a given waypoint name is not enforced unless the `unique` flag is set to `true` for one of the instances of that waypoint name.

## Portals
* **Portals** are defined as `(entrance, exit)` pairs of waypoints.
    * The `entrance` is a bare waypoint name, but the `exit` is a waypoint name paired with an optional **subworld name**.  Omitting the subworld name implies a portal that exits to the same subworld as its entrance.
* Portals can have multiple entrances but only a single exit.
* A waypoint with multiplicity > 1 will result in multiple portals (all to the same exit) when specified as an `entrance`.

### Obstructions
* Portal **entrances** can be blocked, which will prevent movement into the portal.  Portal **exits** are not checked for obstructions when traversing the portal.

# Other musings
* Should some robots not be able to go through portals? e.g. based on capability (c.f. [The Terminator](https://en.wikipedia.org/wiki/The_Terminator) which needs an organic skin  covering)
* We could keep track of how many times a robot has passed through a portal in its internal state. Is there a use for such a statistic?
    * Come to think of it, there are other figures that may also be interesting to track, such as total distance moved.

### CODE REVIEW

**Comment by byorgey** [comment]
Looks fun! I'm travelling but hope to be able to look at this within the next few days sometime.

File: src/Swarm/Game/Step.hs [hunk 1]
```diff
@@ -2677,12 +2688,17 @@ updateRobotLocation ::
...
       robotsByLocation . at oldLoc %= deleteOne rid
-      robotsByLocation . at newLoc . non Empty %= IS.insert rid
-      modify (unsafeSetRobotLocation newLoc)
+      robotsByLocation . at newlocWithPortal . non Empty %= IS.insert rid
+      modify (unsafeSetRobotLocation newlocWithPortal)
       flagRedraw
  where
+  applyPortal loc = do
+    lms <- use worldNavigation
+    return $ M.findWithDefault loc loc $ portals lms
```
Comment (byorgey) [inline src/Swarm/Game/Step.hs:2700]: So am I understanding correctly that if you simply walk on top of a portal, you are instantly teleported?  What's the argument for doing it that way as opposed to, say, having a command like `travel` (or whatever) that activates a portal on which you are standing?  I'm honestly not sure which I would prefer, I can think of pros and cons of both.

File: data/scenarios/Testing/1356-portals/automatic-waypoint-patrol.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+          '@': [dirt, boulder]
+          'w':
+            cell: [dirt]
+            waypoint:
+              name: wp
+        map: |
+          @@@
+          @w.
+          @.@
+  placements:
```
Comment (byorgey) [inline data/scenarios/Testing/1356-portals/automatic-waypoint-patrol.yaml:45]: How do you control the order of all the waypoints in a specific cycle?  Is it determined by the order of the `placements` list?  If so, that seems fragile, especially if the order of placements may also affect which structures get drawn on top of others.

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ where
+  duplicated = M.filter ((> 1) . NE.length) binnedMap
+
+-- | Enforces the following constraints:
+-- * portals can have multiple entrances but only a single exit
+-- * no two portals share the same entrance location
+-- * global waypoint uniqueness when the "unique" flag is specified
+validateNavigation ::
+  (MonadFail m, Traversable t) =>
+  V2 Int32 ->
```
Comment (byorgey) [inline src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:73]: Why is this a `V2 Int32` as opposed to a `Location`?

**Review by byorgey — APPROVED** [review]
Looks cool!  Just a few questions/comments.

File: test/integration/Main.hs [hunk 4]
```diff
@@ -296,6 +296,7 @@ testScenarioSolution _ci _em =
         , testSolution Default "Testing/1234-push-command"
         , testSolution Default "Testing/1256-halt-command"
         , testSolution Default "Testing/1295-density-command"
+        , testSolution Default "Testing/1356-portals/portals-flip-and-rotate.yaml"
```
[the remark this answers is not in the document]
Comment (kostmo) [reply test/integration/Main.hs:299]: This is the only one I defined a goal and solution for; the other two are just "manual" tests.

File: src/Swarm/Game/Step.hs (hunk 1 above)
Comment (kostmo) [reply src/Swarm/Game/Step.hs:2700]: Ultimately I think we should support both :)

Teleporter-style, which is currently implemented, should be instantaneous.  But "ladders" and "stairs" would also be implemented as portals, and we could give the player the option to proceed down or up explicitly, after walking on to the cell.

To specify this, there will be an additional YAML attribute "direction" which can be, `up`, `down`, or omitted.

File: data/scenarios/Testing/1356-portals/automatic-waypoint-patrol.yaml (hunk 2 above)
Comment (kostmo) [reply data/scenarios/Testing/1356-portals/automatic-waypoint-patrol.yaml:45]: Good point.  Perhaps an explicit sort should be performed upon construction.  Opened #1366.

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            [ "No waypoint named"
+            , quote rawName
+            ]
+    Just xs -> return xs
+
+  extractLoc (Originated _ (Waypoint _ loc)) = loc
+  correctedWaypoints =
+    binTuples $
+      map
+        (\x -> (wpName $ wpConfig $ value x, fmap (offsetWaypoint upperLeft) x))
```
Comment (kostmo) [inline src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:116]: Using the `upperLeft` here as a vector to apply an offset

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs (hunk 3 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:73]: I see how that is confusing; I [use it as a vector to offset a location](https://github.com/swarm-game/swarm/pull/1356/files#r1260462962), but really I should pass it into this function as a `Location` and defer the conversion.  Will follow up with fix.

**Comment by kostmo** [comment]
Oops, `mergebot` got stuck so I clicked the merge button.  Should have manually rebased first to pre-empt the breakage.

**Pull request merged**